Implement fair scheduler for spark local mode

xiajunluan opened the pull request:
1. Abstract the TaskSetManager interface for ClusterTaskSetManager and LocalTaskSetManager
2. Use original thread pool for scheduler resource
3. Finish the local end-to-end test for scheduler algorithm

xiajunluan committed:
implement FIFO and fair scheduler for spark local mode

xiajunluan committed:
1. Add unit test for local scheduler
2. Move localTaskSetManager to a new file

AmplabJenkins commented:
Thank you for your pull request. An admin will review this request soon.

mateiz commented:
Jenkins, this is ok to test.

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75/

mateiz commented on core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
```diff
@@ -0,0 +1,734 @@
+        NODE_LOCAL
+      }
+    }
+  }
```
Why do we need to allow invalid locality strings? I'd rather not allow them and set various config options, etc to have good defaults. Printing a warning in this case is too risky, users might miss it.

mridulm replied:
This is primarily used to parse system property specified user - so in case
of errors : default to node_local after flagging a warning.

On Sun, Jun 9, 2013 at 6:57 AM, Matei Zaharia notifications@github.comwrote:

> In core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
> 
> > -  def parse(str: String): TaskLocality = {
> > -    // better way to do this ?
> > -    try {
> > -      val retval = TaskLocality.withName(str)
> > -      // Must not specify PROCESS_LOCAL !
> > -      assert (retval != TaskLocality.PROCESS_LOCAL)
> >   +
> > -      retval
> > -    } catch {
> > -      case nEx: NoSuchElementException => {
> > -        logWarning("Invalid task locality specified '" + str + "', defaulting to NODE_LOCAL");
> > -        // default to preserve earlier behavior
> > -        NODE_LOCAL
> > -      }
> > -    }
> > -  }
> 
> Why do we need to allow invalid locality strings? I'd rather not allow
> them and set various config options, etc to have good defaults. Printing a
> warning in this case is too risky, users might miss it.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mesos/spark/pull/632/files#r4603105
> .

mateiz replied:
Ah, I see, this is older code that was moved. I would still prefer to throw an exception here instead of warning, but we can do it in a different patch since this patch doesn't contain the code that actually parses this. Basically my thinking is that if the user has a mistake in their config file, why continue?

mateiz commented on core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:
```diff
@@ -2,19 +2,51 @@ package spark.scheduler.local
+
+  def launchTask(tasks : Seq[TaskDescription]) {
+    for (task <- tasks)
+    {
```
Put this curly brace on the previous line

mateiz commented on core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:
```diff
@@ -30,89 +62,128 @@ private[spark] class LocalScheduler(threads: Int, maxFailures: Int, sc: SparkCon
 
-  override def start() { }
+    localActor = env.actorSystem.actorOf(
+      Props(new LocalActor(this, threads)), "Test")
```
This can fit on one line

mateiz commented on core/src/main/scala/spark/scheduler/local/LocalScheduler.scala:
```diff
@@ -30,89 +62,128 @@ private[spark] class LocalScheduler(threads: Int, maxFailures: Int, sc: SparkCon
+      val sortedTaskSetQueue = rootPool.getSortedTaskSetQueue()
+      for (manager <- sortedTaskSetQueue) {
+        logInfo("parentName:%s,name:%s,runningTasks:%s".format(manager.parent.name, manager.name, manager.runningTasks))
+      }
```
This should probably be logDebug instead of logInfo

mateiz commented on core/src/main/scala/spark/scheduler/local/LocalTaskSetManager.scala:
```diff
@@ -0,0 +1,173 @@
+    if (availableCpus > 0 && numFinished < numTasks) {
+      findTask() match {
+        case Some(index) =>
+          logInfo(taskSet.tasks(index).toString)
```
This looks like a debug message -- you can remove it

xiajunluan commented:
Hi Matei

I have changed code based on your comments.

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90/

mateiz commented on core/src/test/scala/spark/scheduler/LocalSchedulerSuite.scala:
```diff
@@ -0,0 +1,171 @@
+          assert(ans.toList === List(threadIndex))
+          sem.release()
+          TaskThreadInfo.threadToRunning(threadIndex) = false
+        }
```
Indentation here seems wrong -- should be just two spaces

mateiz commented on core/src/test/scala/spark/scheduler/LocalSchedulerSuite.scala:
```diff
@@ -0,0 +1,171 @@
+          TaskThreadInfo.threadToRunning(threadIndex) = false
+        }
+    }.start()
+    Thread.sleep(2000)
```
Is this to wait for the thread to get started? It would be better to use a condition variable or semaphore instead of sleeping, because this is a long time to sleep in a unit test.

mateiz commented on core/src/test/scala/spark/scheduler/LocalSchedulerSuite.scala:
```diff
@@ -0,0 +1,171 @@
+    assert(TaskThreadInfo.threadToRunning(6) === false)
+    
+    TaskThreadInfo.threadToLock(1).jobFinished()
+    Thread.sleep(1000)
```
This sleep would also be nice to replace with some kind of semaphore or shared state; at the very least have a loop that checks the `ThreadInfo.threadFinished` variable every 100ms, and make that a ConcurrentMap.

ryanlecompte replied:
You could also just use a simple java.util.concurrent.CountDownLatch with the proper size. 

On Jun 9, 2013, at 3:33 PM, Matei Zaharia notifications@github.com wrote:

> In core/src/test/scala/spark/scheduler/LocalSchedulerSuite.scala:
> 
> > +
> > -    createThread(1,null,sc,sem)
> > -    createThread(2,null,sc,sem)
> > -    createThread(3,null,sc,sem)
> > -    createThread(4,null,sc,sem)
> > -    createThread(5,null,sc,sem)
> > -    createThread(6,null,sc,sem)
> > -    assert(TaskThreadInfo.threadToRunning(1) === true)
> > -    assert(TaskThreadInfo.threadToRunning(2) === true)
> > -    assert(TaskThreadInfo.threadToRunning(3) === true)
> > -    assert(TaskThreadInfo.threadToRunning(4) === true)
> > -    assert(TaskThreadInfo.threadToRunning(5) === false)
> > -    assert(TaskThreadInfo.threadToRunning(6) === false)
> > -   
> > -    TaskThreadInfo.threadToLock(1).jobFinished()
> > -    Thread.sleep(1000)
> >   This sleep would also be nice to replace with some kind of semaphore or shared state; at the very least have a static volatile threadsFinished variable (in some kind of static object) and do while (threadsFinished < 4) { Thread.sleep(100) }
> 
> —
> Reply to this email directly or view it on GitHub.

mateiz commented on core/src/test/scala/spark/scheduler/LocalSchedulerSuite.scala:
```diff
@@ -0,0 +1,171 @@
+}
+
+
+class LocalSchedulerSuite extends FunSuite with LocalSparkContext {
```
This test could use a little bit of documentation at the top just to explain how it works. It doesn't need to be long, but just say that we'll launch one task per job and have the task's thread wait for a the test to signal it that it can finish (through TaskThreadInfo).

mateiz commented:
Hey Andrew,

Thanks for fixing those. I made a few other comments on the unit tests -- in particular, I'd really like to avoid having to sleep(2000) a bunch of times in there, because it will make the test take a long time. Other than that, this looks really good. Let me know when you've updated it!

xiajunluan committed:
change code style and debug info

rxin commented:
Hi Andrew - just want to make sure you are aware that there are a couple more comments Matei/Ryan made on the code section, which is not displayed in the main pull request page here (kind of confusing).

xiajunluan commented:
Hi Reynold - I have got it, and will update the code after more tests. Thanks for reminder.

xiajunluan committed:
Update LocalSchedulerSuite to avoid using sleep for task launch

xiajunluan commented:
Hi Matei

I have added completion variable for thread to avoid sleep(2000) to wait job start, but for tasks that will pend for scheduling, I could not think out better solution other than sleep to wait stage to be added to taskSetManager queue for later scheduler when cluster has free resource, I have added some comments in codes for this scenario, pls help me to review codes and call me if there is better solution.

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103/

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104/

mateiz commented:
Hi Andrew,

The remaining sleeps seem fine, since there are very few of them. Thanks! I've merged this manually because your branch has diverged a bit from master and was missing two changes in TaskSetManager.
